File: java/Context.java
import java.net.InetAddress;
import java.lang.management.ManagementFactory;
import java.text.DecimalFormat;

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;

import java.net.URL;
import java.net.MalformedURLException;

class Context {

  //----------------------------------------------------------------
  // Private Variables
  //----------------------------------------------------------------

  private static DecimalFormat  time_formater = new DecimalFormat("0.000000");
  private static double         nano_offset   = System.currentTimeMillis()/1000D - System.nanoTime()/1000000000D;
  private static BufferedWriter log_file      = null;

  //----------------------------------------------------------------
  // Private Methods
  //----------------------------------------------------------------

  private static String getTime() {
    return time_formater.format(nano_offset + (System.nanoTime()/1000000000D));
  }

  private static String getMethodName(final int stack_depth_up) {
    return Thread.currentThread().getStackTrace()[stack_depth_up].getMethodName();
  }

  private static String genMsg(String text, String type) {
    String   time       = getTime();
    String[] pid_string = java.lang.management.ManagementFactory.getRuntimeMXBean().getName().split("@");
    StringBuffer msg    = new StringBuffer();
    msg.append(time);          // time
    msg.append(" ");
    msg.append(pid_string[1]); // host
    msg.append(" ");
    msg.append(pid_string[0]); // pid
    msg.append(" ");
    msg.append(Thread.currentThread().getName().replace(" ","-")); // thread
    msg.append(" ");
    msg.append(type);          // type (START, ENDOK, BMARK, etc)
    msg.append(" ");
    msg.append(getMethodName(4)); // function name
    msg.append(" ");
    msg.append(text);          // text
    msg.append("\n");
    return msg.toString();
  }

  private static void compactStringBuffer(StringBuffer text, Object o) {
      if (o.getClass().isArray()) {
        for (Object oo : (Object[])o) {
          compactStringBuffer(text, oo);
        }
      }
      else {
        text.append(o.toString());
        text.append(" ");
      }
  }
  private static String compactStrings(Object[] objs) {
    StringBuffer text = new StringBuffer();
    compactStringBuffer(text, objs);
    return text.toString();
  }
  
  private static synchronized void logMsg(String msg) {
    try {
      if (log_file != null) {log_file.write(msg);}
      else                  {System.out.print(msg);}
    }
    catch (IOException e) {
      System.err.println("Unable to write to log file " + e.getMessage());
      closeLog();
    }
  }

  //----------------------------------------------------------------
  // Public Methods
  //----------------------------------------------------------------

  public static void setLog(String filename) throws MalformedURLException {
  	setLog(new URL(filename));
  }

  public static void setLog(URL url) {
    try {
		if(url.getProtocol().equals("file")) {
	      log_file = new BufferedWriter(new FileWriter(url.getHost() + url.getPath(), false));
		}
		else {
		}
    }
    catch (IOException e){
      System.err.println("Unable to open log file: " + e.getMessage());
    }
  }
  public static void closeLog() {
      try {log_file.close();} catch (Exception e) {}
      log_file = null;
  }

  public static void start(Object... text) {logMsg(genMsg(compactStrings(text), "START"));}
  public static void endok(Object... text) {logMsg(genMsg(compactStrings(text), "ENDOK"));}
  public static void ender(Object... text) {logMsg(genMsg(compactStrings(text), "ENDER"));}
  public static void bmark(Object... text) {logMsg(genMsg(compactStrings(text), "BMARK"));}
  public static void clear(Object... text) {logMsg(genMsg(compactStrings(text), "CLEAR"));}

}

File: vb/Context.vb
Public Class Context
    Private Shared log_file As System.IO.StreamWriter
    Private Shared epoc As Date = CDate("1.1.1970 00:00:00")
    'Private Shared time_offset As Double
    'Private Shared stopwatch As Stopwatch = setupTimer()

    'Private Shared Function setupTimer() As Stopwatch
    'setupTimer = New System.Diagnostics.Stopwatch()
    'time_offset = DateAndTime.Now.Subtract(epoc).TotalSeconds()
    'setupTimer.Start()
    'End Function

    Private Shared Function getTime() As Double
        getTime = DateAndTime.Now.Subtract(epoc).TotalSeconds
        'getTime = (stopwatch.ElapsedTicks / stopwatch.Frequency) + time_offset
    End Function

    Private Shared Function getMethodName() As String
        Dim s As StackTrace = New System.Diagnostics.StackTrace()
        getMethodName = s.GetFrame(3).GetMethod().Name
    End Function

    Private Shared Function threadName() As String
        threadName = Threading.Thread.CurrentThread.Name
        If Not threadName Is Nothing Or threadName = "" Then
            threadName = CStr(Threading.Thread.CurrentThread.ManagedThreadId)
        End If
    End Function

    Private Shared Function compactString(ByVal o As Object) As String
        compactString = ""
        If o.GetType.IsArray Or o.GetType.IsAssignableFrom(GetType(IEnumerable)) Then
            For Each oo As Object In o
                compactString += compactString(oo)
            Next
        Else
            compactString += o.ToString + " "
        End If
    End Function


    Private Shared Function compactStrings(ByVal ParamArray o() As Object) As String
        compactStrings = compactString(o)
    End Function


    Private Shared Function genMessage(ByVal text As String, ByVal type As String) As String
        genMessage = ""
        genMessage += Format(getTime(), "0.000000") + " " ' time
        genMessage += System.Net.Dns.GetHostName + " " ' host
        genMessage += CStr(Process.GetCurrentProcess.Id) + " " ' PID
        genMessage += threadName().Replace(" ", "-") + " " ' Thread name
        genMessage += type + " "
        genMessage += getMethodName() + " "
        genMessage += text
    End Function

    Private Shared Sub logMsg(ByVal msg As String)
        Try
            If Not log_file Is Nothing Then
                log_file.WriteLine(msg)
            Else
                Debug.Print(msg)
            End If
        Catch ex As Exception
            closeLog()
        End Try
    End Sub

    Public Shared Sub setLog(ByVal filename As String)
        log_file = New System.IO.StreamWriter(filename, False)
    End Sub

    Public Shared Sub closeLog()
        Try
            log_file.Close()
        Catch ex As Exception
        End Try
        log_file = Nothing
    End Sub

    Public Shared Sub start(ByVal ParamArray text() As Object)
        logMsg(genMessage(compactStrings(text), "START"))
    End Sub
    Public Shared Sub endok(ByVal ParamArray text() As Object)
        logMsg(genMessage(compactStrings(text), "ENDOK"))
    End Sub
    Public Shared Sub ender(ByVal ParamArray text() As Object)
        logMsg(genMessage(compactStrings(text), "ENDER"))
    End Sub
    Public Shared Sub bmark(ByVal ParamArray text() As Object)
        logMsg(genMessage(compactStrings(text), "BMARK"))
    End Sub
    Public Shared Sub clear(ByVal ParamArray text() As Object)
        logMsg(genMessage(compactStrings(text), "CLEAR"))
    End Sub

End Class
